docs/root/start/sandboxes/dynamic-configuration.rst

```diff
@@ -0,0 +1,175 @@
+    dynamic-configuration_go-control-plane_1  bin/example -debug             Up (healthy)
+    dynamic-configuration_proxy_1             /docker-entrypoint.sh /usr ... Up            0.0.0.0:10000->10000/tcp, 0.0.0.0:19000->19000/tcp
+    dynamic-configuration_service1_1          /bin/echo-server               Up            8080/tcp
+    dynamic-configuration_service2_1          /bin/echo-server               Up            8080/tcp
+
+Step 5: Query the proxy
+***********************
+
+Once the control plane has started, we should be able to make a request to port ``10000``, which will be
+served by ``service1``.
+
+.. code-block:: console
+
+   $ curl http://localhost:10000
+   Request served by service1
+
+   HTTP/1.1 GET /
+
+   Host: service1
+   Accept: */*
+   X-Forwarded-Proto: http
+   X-Request-Id: 1d93050e-f39c-4602-90f8-a124d6e78d26
+   X-Envoy-Expected-Rq-Timeout-Ms: 15000
+   Content-Length: 0
+   User-Agent: curl/7.72.0
+
+Step 5: Dump the proxy config again
+***********************************
+
+If we now ``config_dump`` the ``dynamic_active_clusters`` we should see the endpoint pointing
+to ``service1``
+
+.. code-block:: console
+
+   $ curl -s http://localhost:19000/config_dump  | jq '.configs[1].dynamic_active_clusters'
+
+.. literalinclude:: _include/dynamic-configuration/response-config-active-clusters.json
+   :language: json
+
+Step 5: Stop the go-control-plane
+*********************************
+
+Let's stop the go-control-plane:
+
+.. code-block:: console
+
+    $ docker-compose stop go-control-plane
+
+The Envoy proxy should continue proxying responses from ``service1``
+
+.. code-block:: console
+
+   $ curl http://localhost:10000 | grep "served by"
+   Request served by service1
+
+Step 5: Edit resource.go and restart the go-control-plane
+*********************************************************
+
+The example go-control-plane is started with a custom resource.go file which
+specifies the configuration provided to Envoy.
+
+If you edit this file and change the ``UpstreamHost`` from ``service1`` to ``service2``:
+
+.. literalinclude:: _include/dynamic-configuration/resource.go
+   :language: go
+   :lines: 33-40
+   :emphasize-lines: 6
+   :linenos:
+
+Further down in this file you must also change the configuration snapshot
+version number from ``"1"`` to ``"2"``:
+
+.. literalinclude:: _include/dynamic-configuration/resource.go
+   :language: go
+   :lineno-start: 167
+   :lines: 167-177
+   :emphasize-lines: 3
+   :linenos:
+
+Now rebuild and restart the go-control-plane:
+
+.. code-block:: console
+
+    $ docker-compose up --build -d go-control-plane
+
+You may need to wait a moment or two for the go-control-plane to become ``healthy``.
+
+Step 5: Query the proxy
+***********************
+
+When you make
+
+.. code-block:: console
+
+   $ curl http://localhost:10000
+   Request served by service2
+
+   HTTP/1.1 GET /
+
+   Host: service1
```

Comment:
this seems like an Envoy bug (maybe im wrong)

basically if you configure a backend  - in this case `service` and then change the configuration dynamically - ie to `service2`, Envoy starts using the correct backend (ie ip) but addresses the new backend with the old backend host name (`service1`)

@dio any ideas about this ?

Reply:
im wondering if this is related to #13659

Reply:
@mattklein123 any ideas about this ?

im wondering if i should open a ticket for it - or if its just a dupe of the ticket i posted

Reply:
How is host being set? Are you using auto host rewrite?  I'm confused who is setting host here. Also, it might be better to use static clusters for this example vs. DNS clusters?

Reply:
How is host being set? 

afaict host is being set by Envoy - its  the only thing that knows the names `service1/2` (apart from docker)

Are you using auto host rewrite? 

not sure - defaults from go-control-plane. You can  see the `cluster` config here 

https://storage.googleapis.com/envoy-pr/13746/docs/start/sandboxes/dynamic-configuration.html#step-4-check-initial-config-and-web-response

altho, im wondering if this is in `listeners`

i tried to override it by setting the host header but didnt work -  got same result

> Also, it might be better to use static clusters for this example vs. DNS clusters?

you mean with fixed IP ?

i think its good to show it like this with dns coming from docker etc (thats how most others do i think) - also, if it is fixed by ip we would have to fix that in docker (maybe i misunderstand)

either way - afaict this behaviour/bug is wrong/undesirable even if we fix the example

im wondering if its the way go-control-plane is setting/updating the config - ill dump more and try to see...

Reply:
after further investigation this seems to be an issue with go-control-plane - at least with the example setup

i ran through the sandbox, and then dumped config...

```console
$ curl -s http://localhost:19000/config_dump  | jq '.configs[4]' 
{
  "@type": "type.googleapis.com/envoy.admin.v3.RoutesConfigDump",
  "dynamic_route_configs": [
    {
      "version_info": "1",
      "route_config": {
        "@type": "type.googleapis.com/envoy.config.route.v3.RouteConfiguration",
        "name": "local_route",
        "virtual_hosts": [
          {
            "name": "local_service",
            "domains": [
              "*"
            ],
            "routes": [
              {
                "match": {
                  "prefix": "/"
                },
                "route": {
                  "cluster": "example_proxy_cluster",
                  "host_rewrite_literal": "service1"
                }
              }
            ]
          }
        ]
      },
      "last_updated": "2020-10-27T19:06:56.600Z"
    }
  ]
}
```

it seems like updating something in go-control-plane is not updating `host_rewrite_literal` when the upstream gets changed

its not an issue with envoy - so sorry for noise

Reply:
it seems like its not updating routes config after the first snapshot, as that is still showing `version=1` even though the rest of the config has `2`

Reply:
im gonna set this as resolved - if its an issue, its outside the scope of this PR

Reply:
Can you just remove ""host_rewrite_literal" ? It's not really needed for the example?

/wait

Reply:
> Can you just remove ""host_rewrite_literal" ? It's not really needed for the example?

not without hacking go-control-plane

Reply:
Go control plane definitely does not force you to set this route config. How are you feeding the config into GCP?

Reply:
cool - i that seems to have worked - updated